File: JavaImpl/src/main/java/org/example/BigONotation/MergeSortExample.java
package org.example.BigONotation;

import java.util.Arrays;

public class MergeSortExample {
    public static void main(String[] args) {
        int[] array = {5, 3, 8, 1, 2};
        mergeSort(array);
        System.out.println("Sorted: " + Arrays.toString(array)); // O(n log n)
    }

    public static void mergeSort(int[] array) {
        if (array.length < 2) return;

        int mid = array.length / 2;
        int[] left = Arrays.copyOfRange(array, 0, mid);
        int[] right = Arrays.copyOfRange(array, mid, array.length);

        mergeSort(left);
        mergeSort(right);
        merge(array, left, right);
    }

    public static void merge(int[] array, int[] left, int[] right) {
        int i = 0, j = 0, k = 0;
        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]) {
                array[k++] = left[i++];
            } else {
                array[k++] = right[j++];
            }
        }
        while (i < left.length) array[k++] = left[i++];
        while (j < right.length) array[k++] = right[j++];
    }
}

File: KotlinImpl/src/main/kotlin/BigONotation/MergeSortExample.kt
package org.example.BigONotation

fun main() {
    val array = intArrayOf(5, 3, 8, 1, 2)
    mergeSort(array)
    println("Sorted: ${array.joinToString(", ")}") // O(n log n)
}

fun mergeSort(array: IntArray) {
    if (array.size < 2) return

    val mid = array.size / 2
    val left = array.copyOfRange(0, mid)
    val right = array.copyOfRange(mid, array.size)

    mergeSort(left)
    mergeSort(right)
    merge(array, left, right)
}

fun merge(array: IntArray, left: IntArray, right: IntArray) {
    var i = 0
    var j = 0
    var k = 0
    while (i < left.size && j < right.size) {
        if (left[i] <= right[j]) {
            array[k++] = left[i++]
        } else {
            array[k++] = right[j++]
        }
    }
    while (i < left.size) array[k++] = left[i++]
    while (j < right.size) array[k++] = right[j++]
}
